Lambda-encoding booleans

robrix opened the pull request:
First blog post in a series about λ encodings.

robrix committed:
Add a draft of a post about lambda-encoded booleans.

robrix committed:
Add the date.

robrix committed:
Add categories.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+# All you need is λ, part one: booleans
+
+Nearly a hundred years ago, Alonzo Church invented computation in the form of the simple, elegant, and utterly inscrutable lambda calculus. Along with Alan Turing, he then proved that anything computable can be computed in the lambda calculus. However, nearly as soon as we had digital computers, we started inventing programming languages, and with them a vast treasure of features, beautiful and t ...
```
Rather than

> Along with Alan Turing…

which could imply that they worked together on the lambda calculus, which isn't quite true, I would say

> Just as Alan Turing proved that anything computable can be computed with a Turing machine, so did Church with the lambda calculus…

or something along those lines. You could mention the Church-Turing thesis as well, if you wanted.

I would also drop the `utterly inscrutable`, since I don't think that's true! It's just unfamiliar to most people.

robrix replied:
The proof that the lambda calculus is Turing-complete—Church-Turing—is precisely what that sentence is describing.

I stand by my description of the lambda calculus as utterly inscrutable; in the absence of a metalanguage, it’s like reading binary machine code or something.

patrickt replied:
Does the Church-Turing thesis technically have a proof, since it's an informal statement? Anyway, I thought it would be nice to call it out by name, but your call.

And fair enough. It cannot be said to leap off the page, though I'd say it's easier than x86 assembly. 😆

aymannadeem replied:
maybe `elusive` instead of `utterly inscrutable`? I do think unfamiliarity with notation makes it challenging to decipher.

numist replied:
```
Along with Alan Turing, he then proved that anything computable can be computed in the lambda calculus.
```

Fwiw I also think the section would benefit from a few rounds of refinement. I found the redundancy of the `compute` root in this sentence distracting.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+## Unconditional λ
+
+The lambda calculus’ austerity is extreme: you don’t even have booleans. All you have are:
```
```suggestion
The lambda calculus’s austerity is extreme: you don’t even have booleans. All you have are:
```

robrix replied:
This is intentional.

patrickt replied:
Strictly speaking, it's incorrect; suffixed `'` plurals only apply if the word itself is plural _and_ ends in an `s`, e.g. it's `bus's`, `mice's`, but `cheerleaders'` (see _Dreyer's English_, p79). It's not the end of the world, though.

numist replied:
`calculi's`

patrickt replied:
That would work too.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+    The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (tho less idiomatic) equivalent for the latter would be `a(b)(c)`.
+
+3. Variables (written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit, since the paucity of syntax means there’s little risk of ambiguity).
```
This parenthetical is so long that I'd say you should nix it and let the sentences within it flourish. You have a lovely prose style; let it flow.

aymannadeem replied:
```suggestion
Variables. These are written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit since the paucity of syntax means there’s little risk of ambiguity.
```

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+How do you _do_ anything when you don’t even have `true` and `false`? Lambdas and variables don’t _do_, they merely _are_, so that leaves application. When all you have is lambda application, everything looks like a lambda abstraction, so we’ll represent booleans using lambdas.
+
+> Aside: Due to the lambda calculus’ terseness, I will be making free use of several notational conveniences:
```
I wouldn't necessarily foist this list into a blockquote, but I'll reserve judgment until I see it rendered on the page.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+2. Function application (“lambda application,” or “application” or “function calls”).
+
+    Applications are written as `x y`, for expressions x and y, and left-associated, i.e. `a b c` = `(a b) c` ≠ `a (b c)`. Function application binds tighter than lambda abstraction, i.e. `λ x . λ y . y x` = `λ x . λ y . (y x)` ≠ `λ x . (λ y . y) x`.
```
Maybe a table comparing LC/Haskell/JS notation, or perhaps a side-by-side implementation of const and id, would be appropriate? Your call.

robrix replied:
`const` and `id` don’t entail application, which is a shame. `fix` is probably too arcane. Hmm.

I was also thinking of maybe giving corresponding implementations for the `Bool` stuff in Haskell (& perhaps JS) but maybe that’s better left to the next post?

patrickt replied:
I think that table, placed near the end of the piece as an act of summarization, would be good.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+    Applications are written as `x y`, for expressions x and y, and left-associated, i.e. `a b c` = `(a b) c` ≠ `a (b c)`. Function application binds tighter than lambda abstraction, i.e. `λ x . λ y . y x` = `λ x . λ y . (y x)` ≠ `λ x . (λ y . y) x`.
+
+    The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (tho less idiomatic) equivalent for the latter would be `a(b)(c)`.
```
```suggestion
    The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (though less idiomatic) equivalent for the latter would be `a(b)(c)`.
```

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+> 5. writing type signatures as tho we had a type system, and even a typechecker, with as much polymorphism and inference as is convenient at any particular moment.
+> 6. using syntactic recursion as tho it existed.
+> 7. using general recursion as tho it made sense.
+> 8. ignoring application order, normalization, reduction, substitution, values, references, allocation, copying, space, time, entropy, and any and all other such details whenever I feel like it.
```
`s/tho/though` here (English orthography is a fuck but it's our unhappy burden to bear)

robrix replied:
I dunno, man, language evolves. We don’t write “although” as much as we used to, for example.

patrickt replied:
I think a full `though` fits the tone of this better; it's mostly-formal writing, after all. It somewhat interrupted the flow for me; language indeed evolves, but formal writing is not the best petri dish.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+>
+> Further, I will by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.
+>
+> I will try to avoid pulling rabbits from hats too wantonly, but for now, I’ll ask you to suspend disbelief; I hope to revisit and justify some of these in later posts.
```
I don't think you necessarily need this disclaimer. The above syntactic sugar is hardly excessive.

robrix replied:
I mostly want to indicate that I intend to justify even more reduction of syntax into the lambda calculus.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+(Note: there aren’t any keywords in the lambda calculus, so there’s nothing stopping me from naming variables things like `if`, a fact which I will take free advantage of.)
+
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
```
```suggestion
We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’s stark palette makes it easy to enumerate _all_ the things we can do with a variable:
```

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
+
+1. Ignore it, or shadow it with another lambda binding the same name.
```
```suggestion
1. Ignore it, or shadow it with another lambda binding of the same name.
```

robrix replied:
Binding is used as a verb here, not a noun.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
+
+1. Ignore it, or shadow it with another lambda binding the same name.
+2. Mention it (perhaps nested), whether on its own in the body of a lambda, or somewhere on either side of an application (or both).
```
Can you expand on precisely what you mean by `(perhaps nested)`?

aymannadeem replied:
Example?

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+We didn’t move the problem around at all—we actually solved it (once we defined `true` and `false`, anyway). It’s simply that encoding booleans in this manner makes `if` largely pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
+
+> Aside: we chose to define `if` as applying the `Bool` to the other arguments in the same order it received them, but we could just as easily have swapped them:
```
If you have footnote support in your blog, these asides would be excellent as footnotes (or even better, sidenotes a là Edward Tufte, but that's a bit fiddly to get working).

robrix replied:
I’m actually pretty 👎 on footnotes, but I was thinking of literally pulling them into an adjacent column, or collapsing them using `<detail>`.

patrickt replied:
Adjacent column would be fine. It's what I do on my blarg.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
````diff
@@ -0,0 +1,260 @@
+false = λ then else . else
+```
+
+We didn’t move the problem around at all—we actually solved it (once we defined `true` and `false`, anyway). It’s simply that encoding booleans in this manner makes `if` largely pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
````
```suggestion
We didn’t move the problem around at all—we solved it (once we defined `true` and `false`, anyway). The fact is that encoding booleans in this manner makes `if` pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
```

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+It’s frequently convenient to conflate booleans with bits, their minimal representation, but in truth they’re not the same at all. Practically, some programming languages define booleans as a byte in memory, perhaps clamping its values to 0 and 1; others define them as instances of some boolean class, or constructors of an algebraic datatype. Some provide no formal relationship between `true` and ...
+
+Mathematically, booleans are the values in propositional logic; the upper and lower bounds of a lattice; the zero and one of a semiring; the members of the set with cardinality 2; and many other things in many different contexts.
+
```
I would emphasize `representing it with functions which can perform all the operations possible with the original datatype`, since it's the linchpin of this piece.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
````diff
@@ -0,0 +1,260 @@
+not = λ x . if x ? ?
+```
+
+> Aside: as discussed in a previous aside, `if` is operationally redundant—i.e. `if x y z` is operationally equivalent to `x y z`—given the ordering of arguments to `Bool`s which we selected earlier. It is, however, pleasantly evocative, and so is used for clarity and so we can stop talking about that ordering decision.
````
I never realized this. So cool!

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
+
+1. Ignore it, or shadow it with another lambda binding the same name.
```
Fair!

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+## Unconditional λ
+
+The lambda calculus’ austerity is extreme: you don’t even have booleans. All you have are:
```
This is a personal style preference for readability that would feel redundant to some, but I generally prefer numerated/bulleted ideas introduced before their explanations are introduced. This gives my brain a primer of 1) how many things I'm about to read about and 2) what their names are _before_ I learn about them. 

Something like:

> All you have are [Single-argument functions](), [Function application](), [Variables]()

followed by longer definitions/explanations or:


> All you have are:
> 1. [Single-argument functions]()
> 2. [Function application](URL)
> 3. [Variables](url)

and then their longer explanations.

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+1. Single-argument functions (“lambda abstractions,” but “lambdas,” “abstractions,” and “functions” will also be used interchangeably).
+
+    These are written `λ x . y`, for variable `x` and expression `y`, where `x` is now available as a bound variable in the body, and any enclosing definition of `x` is shadowed. (We shall assume strictly lexical scoping for the time being.)
```
> and any enclosing definition of `x` is shadowed

maybe provide a small example in parens?

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+    These are written `λ x . y`, for variable `x` and expression `y`, where `x` is now available as a bound variable in the body, and any enclosing definition of `x` is shadowed. (We shall assume strictly lexical scoping for the time being.)
+
+    In Haskell, we would write `\ x -> y` instead; in JavaScript, `function (x) { return y }` or `(x) => y`.
```
👏🏼✨love examples that cross language ecosystem boundaries

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+3. Variables (written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit, since the paucity of syntax means there’s little risk of ambiguity).
+
+Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library—we can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, & values of any variables it closed over when made (again, we assume strict lexical scoping).
```
This sentence is long and a little hard to read. Perhaps consider splitting it into two sentences:

```suggestion
Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library. We can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, and values of any variables it closed over when made (again, we assume strict lexical scoping).
```

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+> 7. using general recursion as tho it made sense.
+> 8. ignoring application order, normalization, reduction, substitution, values, references, allocation, copying, space, time, entropy, and any and all other such details whenever I feel like it.
+>
+> Further, I will by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.
```
```suggestion
> Further, I will abide by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.
```

robrix replied:
“name” is a verb here; “I will, by convention, name types in …”

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library—we can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, & values of any variables it closed over when made (again, we assume strict lexical scoping).
+
+How do you _do_ anything when you don’t even have `true` and `false`? Lambdas and variables don’t _do_, they merely _are_, so that leaves application. When all you have is lambda application, everything looks like a lambda abstraction, so we’ll represent booleans using lambdas.
```
This is an important paragraph since it prompts the reader to think and establishes a critical segue. However, it feels a little abrupt and could use a better transition to start it off and make it more harmonious.

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+> I will try to avoid pulling rabbits from hats too wantonly, but for now, I’ll ask you to suspend disbelief; I hope to revisit and justify some of these in later posts.
+
+
+## λ is blind
```
I love these and the title of the post so much.

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
+
+1. Ignore it, or shadow it with another lambda binding the same name.
```
Example?

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
````diff
@@ -0,0 +1,260 @@
+if = λ cond then else . cond then else
+```
+
+> Aside: under standard semantics for the lambda calculus, we could simplify this definition further by “η-reducing” it, noting that `λ x . f x` behaves the same as `f` alone. Nevertheless, we will leave this and likely other definitions in their full, η-long forms for the sake of clarity.
````
Maybe link to a reference defining  eta-reduction?

aymannadeem commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+## Conclusion
+
+In this post, we’ve explored defining a ubiquitous programming language feature—booleans—using nothing more than the spartan trappings of the lambda calculus, and emerged with a language which can express not merely functions and their applications, but also fundamental metaphysical concepts such as truth.
```
I'm a fan of shorter sentences.

```suggestion
In this post, we’ve explored defining a ubiquitous programming language feature—booleans—using nothing more than the spartan trappings of the lambda calculus. We have emerged with a language which can express not merely functions and their applications, but also fundamental metaphysical concepts such as truth.
```

aymannadeem commented:
Beautifully written! 

When I'm writing posts that have some tutorial-like aspect to them, I like running through the following questions:

- **Who is the audience?** Functional programmers? What's their level of programming proficiency? How many languages are they acquainted with? Are they familiar with the idea of lambda calculus prior to reading this blog post?
- **Why are they learning this?** What can the audience do once they're armed with this information? Does it establish a foundation for topics you intend to cover in future blog posts? What else does it empower them to do?

Sometimes answering these questions first, and then re-reading what I've written helps me clarify my own assumptions or focus topics a bit better.

patrickt commented on _posts/2020-03-29-all-you-need-is-lambda-1-booleans.md:
```diff
@@ -0,0 +1,260 @@
+
+> Aside: under standard semantics for the lambda calculus, we could simplify this definition further by “η-reducing” it, noting that `λ x . f x` behaves the same as `f` alone. Nevertheless, we will leave this and likely other definitions in their full, η-long forms for the sake of clarity.
+
+This feels strangely like cheating: surely we’ve only moved the problem around. Now instead of `if` making the decision about which argument to return, we’ve deferred it to `cond`. But `if` and `cond` aren’t the same kind of thing, semantically; `if` takes a boolean and two other arguments and returns one of the latter, while `cond` _is_ a boolean—albeit evidently a boolean represented as a funct ...
```
s/kind of thing//

robrix committed:
Everything I know is wrong.

Apparently this would only have been the correct possessive if the thing were plural.

robrix committed:
Rephrase.

robrix committed:
Split up the variable item.

robrix committed:
Introduce the list items before discussing in detail.

robrix committed:
Review.

robrix committed:
Swap the section titles.

robrix committed:
Add a table comparing the syntax of the lambda calculus, Haskell, & JS.

robrix committed:
Add BNF.

robrix committed:
Spell “though” out in full.

robrix committed:
Nesting is explained in the rest of the sentence.

robrix committed:
Rewording.

robrix committed:
Emphasis.

robrix committed:
Clarify.

robrix committed:
Abbreviate the example of shadowing.

robrix committed:
Split up the values paragraph.

robrix committed:
Clarify wording.

robrix committed:
Move the syntactic aside into the syntax section.

robrix committed:
Emphasize the lack of other values.

robrix committed:
Add a segue sentence.

robrix committed:
Examples of ignoring/shadowing variables.

robrix committed:
Examples for mentioning variables.

robrix committed:
Unify the examples slightly.

robrix committed:
Link to a reference for eta-reduction.

robrix committed:
Shorter sentence.

robrix committed:
Simplify wording.

robrix committed:
Further introduction.

robrix committed:
Use <aside> for the asides.

robrix committed:
Ignore the jekyll cache.

robrix committed:
Apparently we need this.

robrix committed:
The heading is already included.

robrix committed:
Merge branch 'develop' into lambda-encoding-booleans

robrix committed:
Add a marker for the summary.

robrix committed:
Fix the categories.

robrix committed:
Place the BNF in a <figure>.

robrix committed:
Fix up the table.

robrix committed:
Spacing.

robrix committed:
Reformat the build-depends field.

robrix committed:
Symlink docs to point at _site.

This should help pages deployments? I hope?

robrix committed:
Commit _site.

robrix committed:
:fire: the docs symlink.

robrix committed:
Rename _site to docs.

robrix committed:
Explicitly pass the default config.

robrix committed:
Set the destination directory to docs.
